### ISSUE TITLE
Help with crating child class containing Matplotlib plots 

### ISSUE THREAD

**Issue by GiuseppeCiccone96** [hanjinliu/magic-class#10]
Hi, 

I am trying to create a child class called PlotData (groupbox) containing 2 plots. This is my code:

```
#Child
	@magicclass(name="Plots",layout="horizontal",widget_type="groupbox")
	class PlotData: 
		#plot widgets in child class 
		plt = field(Figure, options={"nrows": 1, "ncols": 1})
		plt1 = field(Figure, options={"nrows": 1, "ncols": 1})
		def plot_storage_modulus(self):
			self.plt.axes[0].set_yscale('log')
			self.plt.axes[0].set_xscale('log')
			for i in range(self.num_tests-1):
				#set log scale
				self.plt.axes[0].plot(self.data['strain [%]'][i], self.data['storage modulus [Pa]'][i], '-s')
				self.plt.draw()
		
		def plot_loss_modulus(self):
			self.plt1.axes[0].set_yscale('log')
			self.plt1.axes[0].set_xscale('log')
			for i in range(self.num_tests-1):
				#set log scale
				self.plt1.axes[0].plot(self.data['strain [%]'][i], self.data['loss modulus [Pa]'][i], '-o')
				self.plt1.draw()
	

	@PlotData.wraps
	def plot_storage_modulus(self):
	      self.plt.axes[0].set_yscale('log')
	      self.plt.axes[0].set_xscale('log')
	      for i in range(self.num_tests-1):
		      #set log scale
		      self.plt.axes[0].plot(self.data['strain [%]'][i], self.data['storage modulus [Pa]'][i], '-s')
		      self.plt.draw()

	@PlotData.wraps
	def plot_loss_modulus(self):
		self.plt1.axes[0].set_yscale('log')
		self.plt1.axes[0].set_xscale('log')
		for i in range(self.num_tests-1):
			#set log scale
			self.plt1.axes[0].plot(self.data['strain [%]'][i], self.data['loss modulus [Pa]'][i], '-o')
			self.plt1.draw()
```

however I get an error message saying that the "ParentClass has no attribute plt". This makes sense as the plt field was defined inside the child class PlotData. How can I avert this issue? If I define the plf field in the parent class, then 4 plots appear (2 from the parent and 2 from the child). Thanks a lot :)

**Comment by hanjinliu** [comment]
OK, that makes sense :)
In your example, the `plot_storage_modulus` method is defined as that of `BulkRheoGUI`. The `self` in `plot_storage_modulus` is an instance of `BulkRheoGUI` so it has `PlotData` as an attribute. You have to use `self.PlotData.plt` instead.
But it looks like the wrapped functions does not need to have an access to the attributes in `BulkRheoGUI`. You can just define `plot_storage_modulus` in `PlotData` as usual and the code will be simpler (maybe your example code is just a part of your complete code?).

**Comment by GiuseppeCiccone96** [comment]
Hi @hanjinliu , yes the example code is just a small part. 

Probably the simplest thing to do is **not** to use a Child class in this case. I was following this https://hanjinliu.github.io/magic-class/use_wraps.html and the rationale for using the child class PlotData was that I wanted the plots to be confined to the PlotData widget for better appearance... 

Along the same lines, is there a way to make the plots interactive when embedded in the GUI ( ie using plt = field(Figure) )?

**Comment by hanjinliu** [comment]
I think one solution is to define a `property` to make it simple. `@magicclass` ignores `property` objects on construction.

```python
@magicclass(labels="True",popup_mode="popup",widget_type="MainWindow")
class BulkRheoGUI:
    ...
    @property
    def plt(self):
        return self.PlotData.plt

    def plot_storage_modulus(self):
        self.plt.axes[0].set_yscale('log')  # OK!
```

> is there a way to make the plots interactive when embedded in the GUI ( ie using plt = field(Figure) )?
No, currently `matplotlib` widget just supports programmatic plot. `plt = field(Figure)` is identical to your example `plt = field(Figure, options={"nrows": 1, "ncols": 1})` because the options are the default keyword arguments of `Figure`.
To make plots interactive you can use `QtPlotCanvas` in `magicclass.ext.pyqtgraph` if your are familiar with `pyqtgraph`, but it's not well documented yet... (if you are interested in it I can show you some brief examples)
I'll think about making `matplotlib` plot interactive and make it available soon since apparently everyone is more familiar with its API (it'll not be hard because I already have a prototype of interactive canvas in other libraries).

**Comment by GiuseppeCiccone96** [comment]
Thanks a lot for all your help, Hanjin! I think defining a property is a very good idea :) 

It's ok - I will wait for the interactive matplotlib as I am more familiar with its API :)

**Comment by hanjinliu** [comment]
Hi @GiuseppeCiccone96 ,
Interactive canvas is now available in v0.6.1. I'll close this issue but if you have any problem please feel free to open a new one.

**Comment by GiuseppeCiccone96** [comment]
Hi @hanjinliu,
thank you very much for your work! I will give it a try today :)!
